cmd/git/main.go

```diff
@@ -0,0 +1,356 @@
+	pflag.StringVar(&flagValues.target, FlagTarget, "", "The target directory of the clone operation")
+	pflag.StringVar(&flagValues.resultFileCommitSha, FlagResultFileCommitSHA, "", "A file to write the commit sha to.")
+	pflag.StringVar(&flagValues.secretPath, FlagSecretPath, "", "A directory that contains a secret. Either username and password for basic authentication. Or a SSH private key and optionally a known hosts file. Optional.")
+}
+
+func main() {
+	if err := Execute(); err != nil {
+		var exitcode = 1
+		switch err := err.(type) {
+		case *ExitError:
+			exitcode = err.Code
+		}
+
+		os.Exit(exitcode)
+	}
+}
+
+// Execute performs flag parsing, input validation and the Git clone
+func Execute() error {
+	flagValues = settings{}
+	pflag.Parse()
+
+	// create logger and context
+	l := ctxlog.NewLogger("git")
+	ctx := ctxlog.NewParentContext(l)
+
+	err := runGitClone(ctx)
+	if err != nil {
+		ctxlog.Error(ctx, err, "program failed with an error")
+	}
+
+	return err
+}
+
+func runGitClone(ctx context.Context) error {
+	if flagValues.url == "" {
+		return &ExitError{Code: 100, Message: "the 'url' argument must not be empty"}
+	}
+
+	if flagValues.target == "" {
+		return &ExitError{Code: 101, Message: "the 'target' argument must not be empty"}
+	}
+
+	if err := checkEnvironment(ctx); err != nil {
+		return err
+	}
+
+	if err := clone(ctx); err != nil {
+		return err
+	}
+
+	if flagValues.resultFileCommitSha != "" {
+		output, err := git(ctx, "-C", flagValues.target, "rev-parse", "--verify", "HEAD")
+		if err != nil {
+			return err
+		}
+
+		if err := ioutil.WriteFile(flagValues.resultFileCommitSha, []byte(output), 0644); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func checkEnvironment(ctx context.Context) error {
+	var checks = []struct{ toolName, versionArg string }{
+		{toolName: "ssh", versionArg: "-V"},
+		{toolName: "git", versionArg: "version"},
+		{toolName: "git-lfs", versionArg: "version"},
+	}
+
+	for _, check := range checks {
+		path, err := exec.LookPath(check.toolName)
+		if err != nil {
+			return &ExitError{Code: 120, Message: err.Error(), Cause: err}
+		}
+
+		out, err := exec.CommandContext(ctx, path, check.versionArg).CombinedOutput()
+		if err != nil {
+			return err
+		}
+
+		ctxlog.Info(ctx, check.toolName,
+			"path", path,
+			"version", strings.TrimRight(string(out), "\n"),
+		)
+	}
+
+	return nil
+}
+
+func clone(ctx context.Context) error {
+	args := []string{
+		"clone",
+		"--quiet",
+		"--no-tags",
+	}
+
+	var commitSha string
+	switch {
+	case commitShaRegEx.MatchString(flagValues.revision):
+		commitSha = flagValues.revision
+		args = append(args,
+			"--no-checkout",
+		)
+
+	case flagValues.revision != "":
+		args = append(args,
+			"--branch", flagValues.revision,
+			"--depth", "1",
+			"--single-branch",
+		)
+
+	default:
+		args = append(args,
```

Comment:
do you know if this have the same behaviour as doing a symbolic-link to HEAD branch, which afaik is the current Tekton behaviour for getting the default branch.

Reply:
I tested this and the behaviour is the same, so resolving.